schema.cc

```diff
@@ -1618,6 +1643,52 @@ schema_ptr schema::get_reversed() const {
     });
 }
 
+
+gc_clock::time_point schema::get_gc_before(const dht::decorated_key& dk, const gc_clock::time_point& query_time) const {
+    // if mode = timeout    // default option, if user does not specify tombstone_gc options
+    // if mode = disabled   // never gc tombstone
+    // if mode = immediate  // can gc tombstone immediately
+    // if mode = repair     // gc after repair
+    const auto& options = tombstone_gc_options();
+    switch (options.mode()) {
+    case tombstone_gc_mode::timeout:
+        dblog.info("Get gc_before for ks={}, table={}, dk={}, mode=timeout", ks_name(), cf_name(), dk);
+        return saturating_subtract(query_time, gc_grace_seconds());
+    case tombstone_gc_mode::disabled:
+        dblog.info("Get gc_before for ks={}, table={}, dk={}, mode=disabled", ks_name(), cf_name(), dk);
+        return gc_clock::time_point::min();
+    case tombstone_gc_mode::immediate:
+        dblog.info("Get gc_before for ks={}, table={}, dk={}, mode=immediate", ks_name(), cf_name(), dk);
+        return gc_clock::time_point::max();
+    case tombstone_gc_mode::repair:
+        const std::chrono::seconds& propagation_delay = options.propagation_delay_in_seconds();
+        auto gc_before = gc_clock::time_point::min();
+        auto repair_timestamp = gc_clock::time_point::min();
+        if (_raw._repair_history_map) {
+            const auto it = _raw._repair_history_map->map.find(dk.token());
+            if (it == _raw._repair_history_map->map.end()) {
+                gc_before = gc_clock::time_point::min();
+            } else {
+                repair_timestamp = it->second;
+                gc_before = saturating_subtract(repair_timestamp, propagation_delay);
+            }
+        }
+        dblog.info("Get gc_before for ks={}, table={}, dk={}, mode=repair, repair_timestamp={}, propagation_delay={}, gc_before={}",
```

Comment:
This will make garbage collection very slow, since every lookup will end up hitting this large map.

We should have _min and _max that we can consult before hitting the map, and we should also have a way to avoid looking up again if two tokens that are close together are looked up (as happens on partition scans).

Reply:
we already cache it per partition.

 _min / _max for which?

Reply:
We should cache it per range. Perhaps gc_before should return the token range for which it applies.

min/max timestamp. If a tombstone is older than the minimum timestamp, you can throw it away without lookup. If it's younger than the maximum timestamp, we know it won't get collected. But if gc_before returns a token range and the compactor keeps it, those aren't needed.

Reply:
I do not understand why you meant by return the token range.

We currently have:

gc_before (a time point) = get_gc_before(dk, query_time)

You want it to be:

range (the token range in the interval map that the dk belongs to), gc_before (a time point)= get_gc_before(dk, query_time)

Then we cache the result into another map

unordered_map<range, gc_before>

?

Why is this useful? Given a dk, there is no direct way to query the map. Also, there are two maps to query, slower.

Reply:
Again, like we agreed in previous discussion, let's do the optimization later if this is a real problem and worth the optimization. Querying an interval map is not that horrible.

Reply:
get_gc_before(dk) currently returns only a gc time. In addition to that it could return a dht::token M such that  get_gc_before() for all keys > dk and < M has the same gc time. So the caller doesn't have to call get_gc_before() for those keys. The caller only needs to cache M, not the hole map.

Reply:
We can do that because all callers query keys in token order (readers return sorted keys), not in random order.

Reply:
Sure, we can measure the effect and decide based on the result.

Reply:
if the query has locality for the keys, yes, we can do that, it totally depends on the usage of the caller.

We can even

range (the token range in the interval map that the dk belongs to), gc_before (a time point)= get_gc_before(dk, query_time) 

which contains lower bound as well, not just the upper bound M. 

This means the user has more work to add the cache logic though. 

In addition, if the caller is too lazy, we can have the cache inside the query function, e.g., keep the latest 5 ranges in cache.

Reply:
Compaction and scans have 100% key locality.

Single partition reads should call this just once per query.

Reply:
Nice. This is very good property.